sql/schemachanger: implement ALTER TABLE ... ADD PRIMARY KEY

ajwerner opened the pull request:
This commit extends #86071 to support `ALTER TABLE ... ADD PRIMARY KEY` in
implicit transactions in the case that the existing primary key uses the
default `rowid` primary key. It adopts the ability of #86071 to drop the
`rowid` column along the way.

Release justification: minor improvement to new functionality needed to unblock DMS support.

Release note (sql change): When running `ALTER TABLE ... ADD PRIMARY KEY`
or `ALTER TABLE ... ADD CONSTRAINT ... PRIMARY KEY` in a single-statement,
implicit transaction, when no primary key had been added to the table, the
old `rowid` column which had been automatically created as the table's
`PRIMARY KEY` will now be dropped.

ajwerner committed:
sql/schemachanger: implement ALTER TABLE ... ADD PRIMARY KEY

This commit extends #86071 to support `ALTER TABLE ... ADD PRIMARY KEY` in
implicit transactions in the case that the existing primary key uses the
default `rowid` primary key. It adopts the ability of #86071 to drop the
`rowid` column along the way.

Release note (sql change): When running `ALTER TABLE ... ADD PRIMARY KEY`
or `ALTER TABLE ... ADD CONSTRAINT ... PRIMARY KEY` in a single-statement,
implicit transaction, when no primary key had been added to the table, the
old `rowid` column which had been automatically created as the table's
`PRIMARY KEY` will now be dropped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86195)
<!-- Reviewable:end -->

otan commented:
thanks!!!

ajwerner committed:
sql/schemachanger: add ability to run pre-checks on alter table commands

This way we can introspect the AST prior to deciding that we need to go resolve
the descriptor. This allows us to claw back some round-trips in cases where we
do not support the relevant feature.

Along the way, the change also refactors some of the supported statement code
just a tad.

Release note: None

ajwerner commented:
TFTR!

bors r+
